qctrlopencontrols/pyquil/program.py

```diff
@@ -0,0 +1,209 @@
+
+    if len(rabi_rotations.shape) == 1:
+        rabi_rotations = rabi_rotations[np.newaxis, :]
+    if len(azimuthal_angles.shape) == 1:
+        azimuthal_angles = azimuthal_angles[np.newaxis, :]
+    if len(detuning_rotations.shape) == 1:
+        detuning_rotations = detuning_rotations[np.newaxis, :]
+
+    operations = np.vstack((rabi_rotations, azimuthal_angles, detuning_rotations))
+    offsets = dynamic_decoupling_sequence.offsets
+
+    time_covered = 0
+    program = Program()
+    program += Pragma('PRESERVE_BLOCK')
+
+    for operation_idx in range(operations.shape[1]):
+
+        offset_distance = offsets[operation_idx] - time_covered
+
+        if np.isclose(offset_distance, 0.0):
+            offset_distance = 0.0
+
+        if offset_distance < 0:
+            raise ArgumentsValueError("Offsets cannot be placed properly",
+                                      {'sequence_operations': operations})
+
+        if offset_distance > 0:
+            while (time_covered+gate_time) <= offsets[operation_idx]:
+                for qubit in target_qubits:
+                    program += I(qubit)
+                time_covered += gate_time
+
+        rabi_rotation = operations[0, operation_idx]
+        azimuthal_angle = operations[1, operation_idx]
+        x_rotation = rabi_rotation * np.cos(azimuthal_angle)
+        y_rotation = rabi_rotation * np.sin(azimuthal_angle)
+        z_rotation = operations[2, operation_idx]
+
+        rotations = np.array([x_rotation, y_rotation, z_rotation])
+        zero_pulses = np.isclose(rotations, 0.0).astype(np.int)
+        nonzero_pulse_counts = 3 - np.sum(zero_pulses)
+        if nonzero_pulse_counts > 1:
+            raise ArgumentsValueError(
+                'Open Controls support a sequence with one '
+                'valid pulse at any offset. Found sequence '
+                'with multiple rotation operations at an offset.',
+                {'dynamic_decoupling_sequence': str(dynamic_decoupling_sequence),
+                 'offset': dynamic_decoupling_sequence.offsets[operation_idx],
+                 'rabi_rotation': dynamic_decoupling_sequence.rabi_rotations[
+                     operation_idx],
+                 'azimuthal_angle': dynamic_decoupling_sequence.azimuthal_angles[
+                     operation_idx],
+                 'detuning_rotaion': dynamic_decoupling_sequence.detuning_rotations[
+                     operation_idx]}
+            )
+
+        for qubit in target_qubits:
+            if nonzero_pulse_counts == 0:
+                program += I(qubit)
+            else:
+                if not np.isclose(rotations[0], 0.0):
+                    program += RX(rotations[0], qubit)
+                elif not np.isclose(rotations[1], 0.0):
+                    program += RY(rotations[1], qubit)
+                elif not np.isclose(rotations[2], 0.):
+                    program += RZ(rotations[2], qubit)
+
+        if np.isclose(np.sum(rotations), 0.0):
```

Comment:
Instead could we check nonzero_pulse_counts == 0? That would be more consistent with the check above.

Reply:
This and the next one are related. We need to add the gate_time here because the gates will also take time to execute. What is not required is the check `np.isclose(np.sum(rotations), 0.0)` - because even if the desired unitary is Identity it should add to the time.

Reply:
Even better, nice catch.